src/utils/display-observer.js

```diff
@@ -0,0 +1,248 @@
+    this.targetDisplayMap_ = new Map();
+  }
+
+  /** @override */
+  dispose() {
+    this.docObserver_.disconnect();
+    this.viewportObserver_.disconnect();
+    this.visibilityUnlisten_();
+    this.visibilityUnlisten_ = null;
+    this.targetObserverMultimap_.clear();
+    this.targetDisplayMap_.clear();
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function(boolean)} callback
+   */
+  observe(target, callback) {
+    let callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      callbacks = [];
+      this.targetObserverMultimap_.set(target, callbacks);
+      this.docObserver_.observe(target);
+      this.viewportObserver_.observe(target);
+    }
+    if (pushIfNotExist(callbacks, callback)) {
+      if (this.targetDisplayMap_.has(target)) {
+        setTimeout(() => {
+          const display = getDisplay(
+            this.targetDisplayMap_.get(target),
+            this.isDocVisible_
+          );
+          if (display != null) {
+            callCallbackNoInline(callback, display);
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function()} callback
+   */
+  unobserve(target, callback) {
+    const callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      return;
+    }
+    removeItem(callbacks, callback);
+    if (callbacks.length == 0) {
+      this.targetObserverMultimap_.delete(target);
+      this.targetDisplayMap_.delete(target);
+      this.docObserver_.unobserve(target);
+      this.viewportObserver_.unobserve(target);
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    this.targetObserverMultimap_.forEach((callbacks, target) => {
+      const displays = this.targetDisplayMap_.get(target);
+      const oldDisplay = getDisplay(displays, !this.isDocVisible_);
+      const newDisplay = getDisplay(displays, this.isDocVisible_);
+      if (newDisplay != null && newDisplay !== oldDisplay) {
+        for (let k = 0; k < callbacks.length; k++) {
+          callCallbackNoInline(callbacks[k], newDisplay);
+        }
+      }
+    });
+  }
+
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @param {number} observer
+   * @private
+   */
+  observed_(entries, observer) {
+    const seen = new Set();
+    for (let i = entries.length - 1; i >= 0; i--) {
+      const {target, isIntersecting} = entries[i];
+      if (seen.has(target)) {
+        continue;
+      }
+      seen.add(target);
+      const callbacks = this.targetObserverMultimap_.get(target);
+      if (!callbacks) {
+        continue;
+      }
+      let displays = this.targetDisplayMap_.get(target);
+      if (!displays) {
+        displays = [];
+        this.targetDisplayMap_.set(target, displays);
+      }
+      const oldDisplay = getDisplay(displays, this.isDocVisible_);
+      displays[observer] = isIntersecting;
+      const newDisplay = getDisplay(displays, this.isDocVisible_);
```

Comment:
Ditto double `getDisplay` calls.

Reply:
deduping to [this thread](https://github.com/ampproject/amphtml/pull/31976#discussion_r560640598).